README.md
groupitaire
===========

Solitaire for Math 442 HW5


Changelog
Initialised Git directory.

carddeck.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include "solitaire.h"
#include "cardstack.h"
#include "carddeck.h"

//
// STRING CONSTANTS for changing the color of terminal output (ANSI)
//
char red[]       = {0x1b,0x5b,0x33,0x31,0x6d,0}; // set to red 
char black[]     = {0x1b,0x5b,0x33,0x30,0x6d,0}; // set to black
char neutral[]   = {0x1b,0x5b,0x33,0x39,0x6d,0}; // return to the default color

//
// STRING CONSTANTS for output of suit characters (UNICODE)
//
char spades[]   = {0xE2, 0x99, 0xA0, 0};
char hearts[]   = {0xE2, 0x99, 0xA5, 0};
char clubs[]    = {0xE2, 0x99, 0xA3, 0};
char diamonds[] = {0xE2, 0x99, 0xA6, 0};

//
// TABLES of outputs for card display, inputs for card play
//    
char *out_suit[] = {spades,hearts,clubs,diamonds};
char *in_suit_upper[] = {"S","H","C","D"};
char *in_suit_lower[] = {"s","h","c","d"};
char *face_upper[] = {" ","A","2","3","4","5","6","7","8","9","T","J","Q","K"};
char *face_lower[] = {" ","a","2","3","4","5","6","7","8","9","t","j","q","k"};

int isRed(card_t *c) {
  return c->suit % 2 == RED;
}

int isBlack(card_t *c) {
  return c->suit % 2 == BLACK;
}

int isAce(card_t *c) {
  return c->face == 1;
}

int isKing(card_t *c) {
  return c->face == 13;
}

int isOnTop(card_t *c) {
  return top(c->stack) == c;
}

int isUp(card_t *c) {
  return (c->stack->type == LAIN) || (isOnTop(c) && (c->stack->type == DISCARD));
}

int isSuccessorOf(card_t *c1, card_t *c2) {
  return (c1->face == c2->face+1);
}

int suitsDiffer(card_t *c1, card_t *c2) {
  return (isRed(c1) && isBlack(c2)) || (isRed(c2) && isBlack(c1));
}

int isOkOn(card_t *c1, card_t *c2) {
  return suitsDiffer(c1,c2) && isSuccessorOf(c2,c1);
}

void putCardOfCode(int face, int suit) {
  printf("%s", suit % 2 == RED ? red : black);
  printf("%s", out_suit[suit]);
  printf("%s", face_upper[face]);
  printf("%s", neutral);
}

void putCard(card_t *c) {
  printf("%s",c->put);
}

card_t *cardOf(char *c, deck_t *d) {
  int i;
  if (c[0] >= '2' && c[0] <= '9') {
    i = c[0] - '0';
  } else if (c[0] == 'T' || c[0] == 't') {
    i = 10;
  } else if (c[0] == 'J' || c[0] == 'j') {
    i = 11;
  } else if (c[0] == 'Q' || c[0] == 'q') {
    i = 12;
  } else if (c[0] == 'K' || c[0] == 'k') {
    i = 13;
  } else if (c[0] == 'A' || c[0] == 'a') {
    i = 1;
  } else {
    return NULL;
  }
  int j;
  if (c[1] == 'S' || c[1] == 's') {
    j = SPADES;
  } else if (c[1] == 'H' || c[1] == 'h') {
    j = HEARTS;
  } else if (c[1] == 'C' || c[1] == 'c') {
    j = CLUBS;
  } else if (c[1] == 'D' || c[1] == 'd') {
    j = DIAMONDS;
  } else {
    return NULL;
  }
  return &d->cards[j*13+i-1];
}

deck_t *newDeck() {

  // Make a fresh deck of cards.
  deck_t *deck = malloc(sizeof(deck_t));

  int i,s;
  for (i=0, s=0; s <= 3; s++) {
    for (int f=1; f <= 13; f++, i++) {
      deck->cards[i].suit = s;
      deck->cards[i].face = f;
      sprintf(deck->cards[i].put,"%s%s%s%s",s%2 == RED ? red : black, face_upper[f], out_suit[s], neutral);
      sprintf(deck->cards[i].get,"%s%s", face_lower[f], in_suit_lower[s]);
      deck->cards[i].below = NULL;
      deck->cards[i].stack = NULL;
    }
  }
  
  // Keep it sorted.
  for (i = 0; i < 52; i++) {
    deck->order[i] = i;
  }

  return deck;
}

void srand48(long);
double drand48(void);

void shuffle(deck_t *deck) {

  struct timeval tp; 
  gettimeofday(&tp,NULL); 
  srand48(tp.tv_sec);

  // Perform a Knuth shuffle.
  for (int i = 0; i < 52; i++) {

    // Select the next card.
    int r = (int)((52-i) * drand48());
    if (r != 0) {
      int tmp = deck->order[i];
      deck->order[i] = deck->order[i+r];
      deck->order[i+r] = tmp;
    }

  }
}


cardstack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "solitaire.h"
#include "cardstack.h"
#include "carddeck.h"

cardstack_t *new_cardstack(int type) {
  cardstack_t *stack = malloc(sizeof(cardstack_t));
  stack->top = NULL;
  stack->type = type;
  return stack;
}

cardstack_t *new_cardstack_fed_by(int type, cardstack_t *feed) {
  cardstack_t *stack = new_cardstack(type);
  stack->feed = feed;
  return stack;
}

void put_helper(card_t *c) {
  if (c != NULL) {
    if (c->below != NULL) {
      put_helper(c->below);
      printf(" ");
    }
    putCard(c);
  }
}
void put_cardstack(cardstack_t *stack, int reveal) {
  card_t *c = stack->top;
  if (reveal) {
    put_helper(c);
  } else {
    while (c != NULL) {
      printf("*");
      c = c->below;
    }
  }
}

int is_empty(cardstack_t *stack) {
  return stack->top == NULL;
}

void push(card_t *card, cardstack_t *stack) {
  card->below = stack->top;
  stack->top = card;
  card->stack = stack;
}

card_t *pop(cardstack_t *stack) {
  card_t *card = stack->top;
  stack->top = card->below;
  card->below = NULL;
  card->stack = NULL;
  return card;
}

card_t *top(cardstack_t *stack) {
  return stack->top;
}

void feed(cardstack_t *stack) {
  push(pop(stack->feed),stack);
}
  
void move_onto(card_t *card, cardstack_t *dest_stack) {

  // Keep track of the source stack, the stack that holds 
  // the card to be moved.
  //
  cardstack_t *srce_stack = card->stack;

  // Reassign the stack of the cards at and above the card 
  // being moved; reassign them to the destination stack.
  // 
  card_t *top_card = top(srce_stack);
  for (card_t *c = top_card; c != card; c = c->below) {
    c->stack = dest_stack;
  }
  card->stack = dest_stack;

  // Move them onto the destination stack.
  //
  srce_stack->top = card->below;
  card->below = dest_stack->top;
  dest_stack->top = top_card;

  // Perhaps flip the top card of the source stack's 
  // feeder pile onto the source stack.
  if (is_empty(srce_stack) && srce_stack->feed != NULL) {
    feed(srce_stack);
  }
}


solitaire.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include "solitaire.h"
#include "cardstack.h"
#include "carddeck.h"

arena_t *newArena() {
  arena_t *arena = (arena_t *)malloc(sizeof(arena_t));
  for (int s = 0; s < 4; s++) {
    arena->highest[s] = 0;
  }
  return arena;
}

deal_t *newDeal(deck_t *deck) {

  deal_t *deal = (deal_t *)malloc(sizeof(deal_t));

  // Make all the stacks.
  deal->draw = new_cardstack(DRAW);
  deal->discard = new_cardstack_fed_by(DISCARD,deal->draw);
  for (int p = 0; p < 7; p++) {
    deal->hidden[p] = new_cardstack(HIDDEN);
    deal->lain[p] = new_cardstack_fed_by(LAIN,deal->hidden[p]);
  }

  // Build the draw pile.
  for (int i = 0; i < 52; i++) {
    push(&deck->cards[deck->order[i]],deal->draw);
  }

  // Fill each of the hidden piles.
  for (int i = 0; i < 7; i++) {
    // Place first cards, then second cards, then third...
    for (int j = i; j < 7; j++) {
      push(pop(deal->draw),deal->hidden[j]);
    }
  }
  
  // Flip the top card of each pile.
  for (int i = 0; i < 7; i++) {
    feed(deal->lain[i]);
  }

  // Flip up the first card.
  feed(deal->discard);

  return deal;
}

void putArena(arena_t *arena) {
  for (int s=0; s<4; s++) {
    printf("[");
    putCardOfCode(arena->highest[s],s);
    printf("] ");
  }
  printf("\n");
}

void putDeal(deal_t *deal) {
  for (int p = 6; p >= 0; p--) {
    printf("[");
    put_cardstack(deal->hidden[p],FALSE);
    printf("]");
    put_cardstack(deal->lain[p],TRUE);
    printf("\n");
  }
  put_cardstack(deal->discard,TRUE);
  printf("\n");
  printf("[");
  put_cardstack(deal->draw,FALSE);
  printf("]\n");
}

int arenaTake(card_t *card, arena_t *arena) {
  if (arena->highest[card->suit] == card->face-1) {
    arena->highest[card->suit]++;
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

int makeArenaPlay(card_t *card, arena_t *arena) {

  cardstack_t *stack = card->stack;

  if (isOnTop(card) && (arenaTake(card,arena) == SUCCESS)) {
    pop(stack);
    if (stack->type == LAIN && is_empty(stack) && !is_empty(stack->feed)) {
      feed(stack);
    }
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

cardstack_t *findEmptyLainStack(deal_t *deal) {
  for (int p = 0; p < 7; p++) {
    if (is_empty(deal->lain[p])) {
      return deal->lain[p];
    }
  }
  return NULL;
}

int moveKingOntoFree(card_t *king, deal_t *deal) {
  cardstack_t *dest = findEmptyLainStack(deal);
  if (isUp(king) && dest != NULL) {
    move_onto(king,dest);
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

int moveCardOntoAnother(card_t *card, card_t *onto) {
  cardstack_t *dest = onto->stack;
  printf("onto LAIN %d %d\n", dest->type == LAIN, dest->type);
  printf("card LAIN %d %d\n", card->stack->type == LAIN, card->stack->type);
  printf("isUp %d\n",isUp(card));
  printf("isOnTop %d\n",isOnTop(onto));
  printf("isOkOn %d\n",isOkOn(card,onto));
  if ((dest->type == LAIN) && isUp(card) && isOnTop(onto) && isOkOn(card,onto)) {
    move_onto(card,dest);
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

int pullFromDrawPile(deal_t *deal) {
  if (is_empty(deal->draw)) {
    return FAILURE;
  } else {
    feed(deal->discard);
    return SUCCESS;
  }
}

void getNextPlay(play_t *play, deck_t *deck) {
  static char buffer[80];
  static char cmd[80];
  static char from[80];
  static char onto[80];
  fgets(buffer,80,stdin);
  sscanf(buffer,"%s",cmd);
  if (cmd[0] == 'm') {
    // it's a move onto a lain stack
    sscanf(buffer,"%s %s",cmd,from);
    play->from = cardOf(from,deck);
    if (isKing(play->from)) {
      play->type = KING_PLAY;
      play->onto = NULL;
    } else {
      play->type = MOVE_PLAY;
      sscanf(buffer,"%s %s %s",cmd,from,onto);
      play->onto = cardOf(onto,deck);
    }
  } else if (cmd[0] == 'p') {
    // it's a play onto the arena
    play->type = ARENA_PLAY;
    sscanf(buffer,"%s %s",cmd,from);
    play->from = cardOf(from,deck);
    play->onto = NULL;
  } else if (cmd[0] == 'd') {
    // it's a draw
    play->type = DRAW_PLAY;
    play->from = NULL;
    play->onto = NULL;
  }
}
  
void playSolitaire(deck_t *deck) {

  // initialize
  arena_t *arena = newArena();
  deal_t *deal = newDeal(deck);

  play_t play;
  int status = SUCCESS;
  while (1) {
    putArena(arena);
    printf("\n");
    putDeal(deal);
    printf("\nEnter your play: ");
    getNextPlay(&play,deck);
    switch (play.type) {
    case ARENA_PLAY:
      // Play a card into the arena.
      status = makeArenaPlay(play.from,arena);
      break;
    case KING_PLAY:
      // Move a king card onto an empty lain stack.
      status = moveKingOntoFree(play.from,deal);
      break;
    case MOVE_PLAY:
      // Move a card (possibly with cards above it) onto some lain stack. 
      status = moveCardOntoAnother(play.from,play.onto);
      break;
    case DRAW_PLAY:
      // Draws the next card and puts it on top of the discard pile.
      status = pullFromDrawPile(deal);
      break;
    }
    
    if (status == SUCCESS) {
      printf("Done!\n");
    } else {
      printf("That play wasn't made.\n");
    } 

  }
}
  
void srand48(long);
double drand48(void);

int main(int argc, char **argv) {

  unsigned long seed = 0;
  if (argc == 1) {
    struct timeval tp; 
    gettimeofday(&tp,NULL); 
    seed = tp.tv_sec;
  } else {
    seed = atol(argv[1]);
  }
  printf("Shuffling with seed %ld.\n",seed);
  srand48(seed);

  deck_t *deck = newDeck();
  shuffle(deck);
  playSolitaire(deck);
}


solitaire.h
#ifndef _SOLITAIRE_H
#define _SOLITAIRE_H

#define FALSE 0
#define TRUE (!FALSE)

//
// CONSTANTS for the colors of suits
//
#define BLACK 0
#define RED   1

//
// CONSTANTS for marking suits of cards, and stacks of cards
//
#define SPADES    0
#define HEARTS    1
#define CLUBS     2
#define DIAMONDS  3

// 
// CONSTANTS for marking stacks of cards
//
#define DRAW           4
#define DISCARD        5
#define LAIN           6  
#define HIDDEN         7
#define TEMP           (-1)

// 
// ACKNOWLEDGEMENT of a card play action
//
#define FAILURE 0
#define SUCCESS (!FAILURE)

//
// PLAY action type
//
#define ARENA_PLAY 0
#define KING_PLAY  1
#define MOVE_PLAY  2
#define DRAW_PLAY  3

//
// CARD linked list struct (for stacking cards)
//
typedef struct _card_t {
  int player;
  int face;
  int suit;
  char get[3];
  char put[12];
  struct _card_t *below;
  struct _cardstack_t *stack;
} card_t;

//
// DECK struct for representing the full variety of 52 cards
// 
typedef struct _deck_t {
  int player;
  card_t cards[52];
  int order[52];
} deck_t;

//
// CARDSTACK struct for representing a stacked pile of cards, or a layout of played cards
//
typedef struct _cardstack_t {
  card_t *top;
  struct _cardstack_t *feed;
  int type;  
  // type is one of
  //   a SPADES arena pile
  //   a HEARTS arena pile
  //   a CLUBS arena pile
  //   a DIAMONDS arena pile
  //   the DRAW pile
  //   the DISCARD pile
  //   the 7 hidden piles HIDDEN
  //   the 7 lain piles LAIN
} cardstack_t;

//
// SOLITAIRE struct for the play in front of the player
//
typedef struct _deal_t {
  int player;
  deck_t *deck;
  cardstack_t *hidden[7];
  cardstack_t *lain[7];
  cardstack_t *discard;
  cardstack_t *draw;
} deal_t;

//
// ARENA struct for the playing the "ace through king" stacks
// 
typedef struct _arena_t {
  int highest[4];
} arena_t;

//
// PLAY struct for handling solitaire plays
//
typedef struct _play_t {
  int type;
  card_t *from;
  card_t *onto;
} play_t;

#endif


solitaire_s.c
#include <sys/types.h>
#include <sys/uio.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <strings.h>
#include <errno.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include "solitaire.h"
#include "cardstack.h"
#include "carddeck.h"

#define MAXLINE 80
#define MAXCONNECTIONS 1024

typedef struct _client_t {
	int id;
	int connections;
	srtruct sockaddr_in address;
	arena_t arena;
	deck_t deck;
} client_t;

arena_t *newArena() {
  arena_t *arena = (arena_t *)malloc(sizeof(arena_t));
  for (int s = 0; s < 4; s++) {
    arena->highest[s] = 0;
  }
  return arena;
}

deal_t *newDeal(deck_t *deck) {

  deal_t *deal = (deal_t *)malloc(sizeof(deal_t));

  // Make all the stacks.
  deal->draw = new_cardstack(DRAW);
  deal->discard = new_cardstack_fed_by(DISCARD,deal->draw);
  for (int p = 0; p < 7; p++) {
    deal->hidden[p] = new_cardstack(HIDDEN);
    deal->lain[p] = new_cardstack_fed_by(LAIN,deal->hidden[p]);
  }

  // Build the draw pile.
  for (int i = 0; i < 52; i++) {
    push(&deck->cards[deck->order[i]],deal->draw);
  }

  // Fill each of the hidden piles.
  for (int i = 0; i < 7; i++) {
    // Place first cards, then second cards, then third...
    for (int j = i; j < 7; j++) {
      push(pop(deal->draw),deal->hidden[j]);
    }
  }
  
  // Flip the top card of each pile.
  for (int i = 0; i < 7; i++) {
    feed(deal->lain[i]);
  }

  // Flip up the first card.
  feed(deal->discard);

  return deal;
}

void putArena(arena_t *arena) {
  for (int s=0; s<4; s++) {
    printf("[");
    putCardOfCode(arena->highest[s],s);
    printf("] ");
  }
  printf("\n");
}

void putDeal(deal_t *deal) {
  for (int p = 6; p >= 0; p--) {
    printf("[");
    put_cardstack(deal->hidden[p],FALSE);
    printf("]");
    put_cardstack(deal->lain[p],TRUE);
    printf("\n");
  }
  put_cardstack(deal->discard,TRUE);
  printf("\n");
  printf("[");
  put_cardstack(deal->draw,FALSE);
  printf("]\n");
}

int arenaTake(card_t *card, arena_t *arena) {
  if (arena->highest[card->suit] == card->face-1) {
    arena->highest[card->suit]++;
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

int makeArenaPlay(card_t *card, arena_t *arena) {

  cardstack_t *stack = card->stack;

  if (isOnTop(card) && (arenaTake(card,arena) == SUCCESS)) {
    pop(stack);
    if (stack->type == LAIN && is_empty(stack) && !is_empty(stack->feed)) {
      feed(stack);
    }
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

cardstack_t *findEmptyLainStack(deal_t *deal) {
  for (int p = 0; p < 7; p++) {
    if (is_empty(deal->lain[p])) {
      return deal->lain[p];
    }
  }
  return NULL;
}

int moveKingOntoFree(card_t *king, deal_t *deal) {
  cardstack_t *dest = findEmptyLainStack(deal);
  if (isUp(king) && dest != NULL) {
    move_onto(king,dest);
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

int moveCardOntoAnother(card_t *card, card_t *onto) {
  cardstack_t *dest = onto->stack;
  if ((dest->type == LAIN) && isUp(card) && isOnTop(onto) && isOkOn(card,onto)) {
    move_onto(card,dest);
    return SUCCESS;
  } else {
    return FAILURE;
  }
}

int pullFromDrawPile(deal_t *deal) {
  if (is_empty(deal->draw)) {
    return FAILURE;
  } else {
    feed(deal->discard);
    return SUCCESS;
  }
}

void getNextPlay(int stream, play_t *play, deck_t *deck) {
  static char buffer[MAXLINE];
  static char cmd[MAXLINE];
  static char from[MAXLINE];
  static char onto[MAXLINE];
  read(stream,buffer,MAXLINE);
  sscanf(buffer,"%s",cmd);
  if (cmd[0] == 'm') {
    // it's a move onto a lain stack
    sscanf(buffer,"%s %s",cmd,from);
    play->from = cardOf(from,deck);
    if (isKing(play->from)) {
      play->type = KING_PLAY;
      play->onto = NULL;
    } else {
      play->type = MOVE_PLAY;
      sscanf(buffer,"%s %s %s",cmd,from,onto);
      play->onto = cardOf(onto,deck);
    }
  } else if (cmd[0] == 'p') {
    // it's a play onto the arena
    play->type = ARENA_PLAY;
    sscanf(buffer,"%s %s",cmd,from);
    play->from = cardOf(from,deck);
    play->onto = NULL;
  } else if (cmd[0] == 'd') {
    // it's a draw
    play->type = DRAW_PLAY;
    play->from = NULL;
    play->onto = NULL;
  }
}
  
void sendAck(int client, int ack) {
  if (ack == SUCCESS) {
    write(client,"SUCCESS",8);
  } else {
    write(client,"FAILURE",8);
  }
}

void playSolitaire(void *client) {

  // initialize
  deal_t *deal = newDeal(deck);

  play_t play;
  int status = SUCCESS;
  while (1) {
    putArena(arena);
    printf("\n");
    putDeal(deal);
    printf("\nEnter your play: ");
    getNextPlay(client,&play,deck);
    switch (play.type) {
    case ARENA_PLAY:
      // Play a card into the arena.
      status = makeArenaPlay(play.from,arena);
      break;
    case KING_PLAY:
      // Move a king card onto an empty lain stack.
      status = moveKingOntoFree(play.from,deal);
      break;
    case MOVE_PLAY:
      // Move a card (possibly with cards above it) onto some lain stack. 
      status = moveCardOntoAnother(play.from,play.onto);
      break;
    case DRAW_PLAY:
      // Draws the next card and puts it on top of the discard pile.
      status = pullFromDrawPile(deal);
      break;
    }
    
    sendAck(client,status);
  }
}

int acceptClientOn(int socket) {

    socklen_t acceptlen = sizeof(struct sockaddr_in);
    struct sockaddr_in clientaddr;
    int connection;

    if ((connection = accept(socket, (struct sockaddr *)&clientaddr, &acceptlen)) < 0) {
      fprintf(stderr,"ACCEPT failed.\n");
      exit(-1);
    }

    //
    // Report the client that connected.
    //
    struct hostent *hostp;
    if ((hostp = gethostbyaddr((const char *)&clientaddr.sin_addr.s_addr, 
			       sizeof(struct in_addr), AF_INET)) == NULL) {
      fprintf(stderr, "GETHOSTBYADDR failed.");
    }
    
    printf("Accepted connection from client %s (%s)\n", 
	   hostp->h_name, inet_ntoa(clientaddr.sin_addr));
    
    return connection;
}

int initConnection(int port) {
  //
  // Open a socket to listen for client connections.
  //
  int listenfd;
  if ((listenfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
    fprintf(stderr,"SOCKET creation failed.\n");
    exit(-1);
  }
  
  //
  // Build the service's info into a (struct sockaddr_in).
  //
  struct sockaddr_in serveraddr;
  bzero((char *) &serveraddr, sizeof(struct sockaddr_in));
  serveraddr.sin_family = AF_INET; 
  serveraddr.sin_addr.s_addr = htonl(INADDR_ANY); 
  serveraddr.sin_port = htons((unsigned short)port); 

  //
  // Bind that socket to a port.
  //
  if (bind(listenfd, (struct sockaddr *)&serveraddr, sizeof(struct sockaddr)) < 0) {
    fprintf(stderr,"BIND failed.\n");
    exit(-1);
  }

  //
  // Listen for client connections on that socket.
  //
  if (listen(listenfd, MAXCONNECTIONS) < 0) {
    fprintf(stderr,"LISTEN failed.\n");
    exit(-1);
  }
  
  fprintf(stderr,"Solitaire server listening on port %d...\n",port);

  return listenfd;
}
  
void srand48(long);
double drand48(void);

int main(int argc, char **argv) {

  // 
  // Make sure we've been given a port to listen on.
  //
  if (argc != 2) {
    fprintf(stderr, "usage: %s <port>\n", argv[0]);
    exit(0);
  }
  
  arena_t *arena = newArena();
  int listener = initConnection(atoi(argv[1]));
  
  int clients = 0
  while (1) {
  	clients++;
  	
  //
  // Build a client's profile to send to a handler thread
  //
  client_t *client = (client_t *)malloc(sizeof(client_t));
  
  //
  // Accept a connection from a client, get a file descriptor for communicating
  // with the client
  //
  client->id = clients;
  int client = acceptClientOn(listener);
  }
  
  //
  // create a thread to handle the client
  //
  pthread_t tid;
  pthread_create(&tid,NULL,playSolitaire, (void *)client);

  unsigned long seed = 0;
  if (argc == 1) {
    struct timeval tp; 
    gettimeofday(&tp,NULL); 
    seed = tp.tv_sec;
  } else {
    seed = atol(argv[1]);
  }
  printf("Shuffling with seed %ld.\n",seed);
  srand48(seed);


  deck_t *deck = newDeck();
  shuffle(deck);
  playSolitaire(client);
}
}


solitaire_c.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <strings.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/uio.h>
#include <unistd.h>
#include "solitaire.h"
#include "cardstack.h"
#include "carddeck.h"

#define MAXLINE 80

int connectToServer(char *host, int port) {

  //
  // Look up the host's name to get its IP address.
  //
  struct hostent *hp;
  if ((hp = gethostbyname(host)) == NULL) {
    fprintf(stderr,"GETHOSTBYNAME failed.\n");
    exit(-1);
  }

  //
  // Request a socket and get its file descriptor.
  //
  int clientfd;
  if ((clientfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
    fprintf(stderr,"SOCKET creation failed.\n");
    exit(-1);
  }

  //
  // Fill in the host/port info into a (struct sockaddr_in).
  //
  struct sockaddr_in serveraddr;
  bzero((char *) &serveraddr, sizeof(struct sockaddr_in));
  serveraddr.sin_family = AF_INET;
  bcopy((char *) hp->h_addr_list[0], (char *)&serveraddr.sin_addr.s_addr, hp->h_length);
  serveraddr.sin_port = htons(port);

  //
  // Connect to the given host at the given port number.
  //
  if (connect(clientfd, (struct sockaddr *)&serveraddr, sizeof(struct sockaddr_in)) < 0) {
    fprintf(stderr,"CONNECT failed.\n");
    exit(-1);
  }

  unsigned char *ip;
  ip = (unsigned char *)&serveraddr.sin_addr.s_addr;
  printf("Connected to solitaire service at %d.%d.%d.%d.\n",ip[0],ip[1],ip[2],ip[3]);

  return clientfd;
}

int receiveAck(int socket) {
  char buffer[MAXLINE];
  read(socket,buffer,MAXLINE);
  if (strcmp(buffer,"SUCCESS") == 0) {
    return SUCCESS;
  } else if (strcmp(buffer,"FAILURE") == 0) {
    return FAILURE;
  }
  fprintf(stderr,"Recieved unexpected acknowledgement from the server.\n");
  fprintf(stderr,"Message was '%s'.\n",buffer);
  exit(-1);
}


int main(int argc, char **argv) {

  char buffer[MAXLINE];
  
  //
  // Check for host name and port number.
  //
  if (argc != 3) {
    fprintf(stderr,"usage: %s <host> <port>\n", argv[0]);
    exit(0);
  }

  //
  // Connect to the server.
  //
  int server = connectToServer(argv[1],atoi(argv[2]));

  printf("Welcome to the remote SOLITAIRE game for MATH 442.\n\n");
  printf("Here is a key of possible card plays you can make:\n");
  printf("\tplay <face><suit> - throw a card onto an arena pile in the center.\n");
  printf("\tmove k<suit> - move a king onto an empty pile.\n");
  printf("\tmove <face><suit> <face><suit>- move a card, putting it on top of another card.\n");
  printf("\tdraw - draw and overturn the next card, placing it on the discard pile.\n");
  printf("where <face> is one of a,2,3,...,9,t,j,q,k and <suit> is one of d,c,s,h.\n");

  //
  // Send a series of solitaire play commands.
  //
  while (1) {
    printf("\nEnter a card play: ");
    
    fgets(buffer,MAXLINE,stdin);
    write(server,buffer,strlen(buffer)+1);
    if (receiveAck(server)) {
      printf("Done!\n");
    } else {
      printf("That play wasn't made.\n");
    } 
  }

}
